docs/security.rst

```diff
@@ -0,0 +1,216 @@
+   >>> from zope.lifecycleevent import ObjectCreatedEvent
+   >>> event = ObjectCreatedEvent(Folder())
+   >>> len(find_active_subscriptions_for(event.object, event))
+   1
+   >>> find_active_subscriptions_for(event.object, event)
+   [<...Subscription ... to='https://this_domain_does_not_exist' for=IContentContainer when=IObjectCreatedEvent>]
+
+
+Subscription Is Not Applicable By Default
+-----------------------------------------
+
+Next, we need to know if the subscription is :term:`applicable` to the
+data. Unlike before, since we have security constraints in place, the
+subscription is *not* applicable:
+
+.. doctest::
+
+   >>> subscriptions = find_active_subscriptions_for(event.object, event)
+   >>> [subscription.isApplicable(event.object) for subscription in subscriptions]
+   [True]
+
+Wait, wait...what happened there? It turns out that since we don't
+have any defined principal identified by ``some.one``, we use the
+global ``IUnauthenticatedPrincipal``, an anonymous user. In turn, the
+directives executed by loading ``securitypolicy.zcml`` from
+``zope.securitypolicy`` give anonymous users the ``zope.View``
+permission by default. Let's reverse that and check again.
+
+.. doctest::
+
+   >>> from zope.securitypolicy.rolepermission import rolePermissionManager
+   >>> rolePermissionManager.denyPermissionToRole('zope.View', 'zope.Anonymous')
+   >>> [subscription.isApplicable(event.object) for subscription in subscriptions]
+   [False]
+
+Ahh, that's better.
+
+Subscription Applicable Once Principals are Defined
+---------------------------------------------------
+
+To grant access in an expected way, we'll use
+``zope.principalregistry`` to globally define the prinicpal we're
+looking for, as well as globally grant that principal the permissions
+necessary:
+
+   >>> conf_context = xmlconfig.string("""
+   ... <configure
+   ...     xmlns="http://namespaces.zope.org/zope"
+   ...     xmlns:webhooks="http://nextthought.com/ntp/webhooks"
+   ...     >
+   ...   <include package="zope.securitypolicy" file="meta.zcml" />
+   ...   <include package="zope.principalregistry" file="meta.zcml" />
+   ...   <principal
+   ...         id="some.one"
+   ...         title="Some One"
+   ...         login="some.one"
+   ...         password_manager="SHA1"
+   ...         password="40bd001563085fc35165329ea1ff5c5ecbdbbeef"
+   ...         />
+   ...   <grant principal="some.one" permission="zope.View" />
+   ... </configure>
+   ... """)
+
+Now our webhook is :term:`applicable`:
+
+.. doctest::
+
+   >>> [subscription.isApplicable(event.object) for subscription in subscriptions]
+   [True]
+
+Existing Interactions
+---------------------
+
+If there was already an interaction going on (e.g., for the logged in
+user that created the object), the owner of the subscription is added
+to that interaction for purposes of checking permissions. Security
+policies generally only grant access if all participations in the
+interaction have access.
```

Comment:
I'm sure there is a really good reason for checking the permission against the existing interaction's principle _and_ the owner of the subscription but I'm struggling to grasp why this is needed or what this gets us? I'm not suggesting it's not needed, I'm just not getting it on a first (or second pass). Can you elaborate on your thinking here?

Reply:
The general idea is that whenever two (or more) users interact, only the intersection of their privileges apply. This is to prevent privilege escalation issues. In Zope, the most common motivating example is a user calling another user's uploaded Python script, but the same thing could hold when one user uses part of another users template in their own page, etc. 

It can also be used as an additional filtering mechanism.

Here, we have a user making the subscription, and a user generating the event that triggers the subscription. Suppose you want to get notified every time any administrator creates a new `IContentContainer`.

```xml
     <webhooks:staticSubscription
               to="https://this_domain_does_not_exist"
               for="zope.container.interfaces.IContentContainer"
               when="zope.lifecycleevent.interfaces.IObjectCreatedEvent" />
```

That won't do it, you'll get notified for *everything* created by *anyone* because there is no security declaration there.

```xml
     <webhooks:staticSubscription
               to="https://this_domain_does_not_exist"
               for="zope.container.interfaces.IContentContainer"
               when="zope.lifecycleevent.interfaces.IObjectCreatedEvent"
               owner="zope.security.system_user"
               permission="zope.Manage" />
```

That will work. We've declared that this subscription belongs to a user that always exists and always has the admin permission, but because it also takes the currently acting user into account, it will only fire (be "applicable") when they also have that permission.

A security implication is that after this check is run, potentially user-provided code gets to run, in the form of the specified dialect. That code is going to be running with an interaction of the current user[1], but it's going to be running on behalf of the user that owns the subscription. It should only get to run if both users have the necessary permissions.

In practice, for the common types of permissions and objects we're talking about, where the triggering user will typically have (by necessity) the same permission that the subscriber needs, I doubt it will make a difference.

[1] It occurs to me that we might wish to always invoke the dialect in the context of the dual interaction, rather than just doing a check once up front. But…that sounds like a lot of work :)

Reply:
Makes sense, thank you!